aten/src/ATen/ParallelNative.h

```diff
@@ -35,62 +44,43 @@ inline void parallel_for(
   if (begin >= end) {
     return;
   }
-
-  if (((end - begin) >= grain_size) && !in_parallel_region()) {
-    // choose number of tasks based on grain size and number of threads
-    size_t chunk_size = divup((end - begin), get_num_threads());
-    // make sure each task is at least grain_size size
-    chunk_size = std::max((size_t)grain_size, chunk_size);
-    size_t num_tasks = divup((end - begin), chunk_size);
-
-    std::atomic_flag err_flag = ATOMIC_FLAG_INIT;
-    std::exception_ptr eptr;
-    auto task = [f, &eptr, &err_flag]
-        (int64_t task_id, int64_t local_start, int64_t local_end) {
-      internal::_set_thread_num(task_id);
-      internal::_set_in_parallel_region(true);
+  if ((end - begin) < grain_size || in_parallel_region()) {
+    f(begin, end);
+    return;
+  }
+  size_t num_tasks, chunk_size;
+  std::tie(num_tasks, chunk_size) =
+      internal::calc_num_tasks_and_chunk_size(begin, end, grain_size);
+  std::atomic_flag err_flag = ATOMIC_FLAG_INIT;
+  std::exception_ptr eptr;
+  std::vector<std::shared_ptr<c10::ivalue::Future>> futures(num_tasks);
+  for (size_t i = 0; i < num_tasks; ++i) {
+    futures[i] = std::make_shared<c10::ivalue::Future>(NoneType::get());
+  }
+  auto task = [f, &eptr, &err_flag, &futures, begin, end, chunk_size]
+      (int idx, size_t task_id) {
```

Comment:
we don't seem to be using `idx`?

Reply:
We have `idx` because caffe2 thread pool API expects this signature. We could create another lambda wrapper to hide it but I feel a bit overkill... I can rename it to "_" if you prefer.